modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java

```diff
@@ -118,6 +120,17 @@
  * Used when persistence enabled.
  */
 public class GridCacheOffheapManager extends IgniteCacheOffheapManagerImpl implements DbCheckpointListener {
+    /**
+     * Threshold to calculate limit for pages list on-heap caches.
+     * <p>
+     * Note: When a checkpoint is triggered, we need some amount of page memory to store pages list on-heap cache.
+     * If a checkpoint is triggered by "too many dirty pages" reason and pages list cache is rather big, we can get
+     * {@code IgniteOutOfMemoryException}. To prevent this, we can limit the total amount of cached page list buckets,
+     * assuming that checkpoint will be triggered if no more then 3/4 of pages will be marked as dirty (there will be
+     * at least 1/4 of clean pages) and each cached page list bucket can be stored to up to 2 pages.
+     */
+    private static final double PAGE_LIST_CACHE_LIMIT_THRESHOLD = 1.0 / 4.0 / 2.0;
```

Comment:
more exactly this formula is not static, but depends on PagesCache.MAX_SIZE, so if PagesCache.MAX_SIZE > (empty)PagesListNodeIO#getCapacity it can take more than 2 pages.